pkg/plugins/placementpolicy/core/core.go

```diff
@@ -73,38 +57,114 @@ func (m *PlacementPolicyManager) GetPlacementPolicyForPod(ctx context.Context, p
+			filteredPPList = append(filteredPPList, pp)
+		}
+	}
+	return filteredPPList
 }
 
-// AnnotatePod annotates the pod with the placement policy.
-func (m *PlacementPolicyManager) AnnotatePod(ctx context.Context, pod *corev1.Pod, pp *v1alpha1.PlacementPolicy, preferredNodeWithMatchingLabels bool) (*corev1.Pod, error) {
-	annotations := map[string]string{}
-	if pod.Annotations != nil {
-		annotations = pod.Annotations
+func (m *PlacementPolicyManager) GetPolicyInfo(pp *v1alpha1.PlacementPolicy) *PolicyInfo {
+	policyNamespace := pp.Namespace
+	policyName := pp.Name
+	corePolicy := pp.Spec.Policy
+
+	namespacePolicies, namespaceExists := m.policies[policyNamespace]
+
+	if namespaceExists {
+		policy, policyExists := namespacePolicies[policyName]
+
+		if policyExists {
+			return policy
+		}
+
+		created := newPolicyInfo(policyNamespace, policyName, corePolicy.Action, corePolicy.TargetSize)
+		m.policies[policyNamespace][policyName] = created
+		return created
 	}
 
-	preference := "false"
-	if preferredNodeWithMatchingLabels {
-		preference = "true"
+	m.policies[policyNamespace] = make(map[string]*PolicyInfo)
+	created := newPolicyInfo(policyNamespace, policyName, corePolicy.Action, corePolicy.TargetSize)
+	m.policies[policyNamespace][policyName] = created
+	return created
+}
+
+type PodAction int16
+
+const (
+	Match PodAction = iota
+	Add
+	Remove
+)
+
+func (m *PlacementPolicyManager) MatchPod(ctx context.Context, pod *corev1.Pod, policy *PolicyInfo) (*PolicyInfo, error) {
+	matchError := policy.addMatch(pod)
+	if matchError != nil {
+		return nil, matchError
 	}
-	annotations[v1alpha1.PlacementPolicyAnnotationKey] = pp.Name
-	annotations[v1alpha1.PlacementPolicyPreferenceAnnotationKey] = preference
-	pod.Annotations = annotations
-	return m.client.CoreV1().Pods(pod.Namespace).Update(ctx, pod, metav1.UpdateOptions{})
+
+	m.updatePolicies(policy, Match)
+	return policy, nil
 }
 
-func (m *PlacementPolicyManager) GetPlacementPolicy(ctx context.Context, namespace, name string) (*v1alpha1.PlacementPolicy, error) {
-	return m.ppLister.PlacementPolicies(namespace).Get(name)
+func (m *PlacementPolicyManager) AddPod(ctx context.Context, pod *corev1.Pod, policy *PolicyInfo) (*PolicyInfo, error) {
+	addError := policy.addPodIfNotPresent(pod)
+	if addError != nil {
+		return nil, addError
+	}
+
+	m.updatePolicies(policy, Add)
+	return policy, nil
 }
 
-func (m *PlacementPolicyManager) filterPlacementPolicyList(ppList []*v1alpha1.PlacementPolicy, pod *corev1.Pod) []*v1alpha1.PlacementPolicy {
-	var filteredPPList []*v1alpha1.PlacementPolicy
-	for _, pp := range ppList {
-		labels := pp.Spec.PodSelector.MatchLabels
-		if utils.HasMatchingLabels(pod.Labels, labels) {
-			filteredPPList = append(filteredPPList, pp)
+func (m *PlacementPolicyManager) RemovePod(pod *corev1.Pod) error {
+	key, keyError := framework.GetPodKey(pod)
+	if keyError != nil {
+		return keyError
+	}
+
+	podNamespace := pod.Namespace
+	ppList := m.policies[podNamespace]
+
```

Comment:
if `ppList == nil` we can return early here and move the rest of the code out of the if block